proposals/p2240.md

```diff
@@ -588,85 +604,198 @@ rejected), so we will focus on the semantics of phase 3 when the pattern type is
 a variadic type. Variadic types are themselves patterns, so we will break this
 down into cases based on the possible forms of a pattern.
 
-**NOTE TO READER:** You should probably ignore these examples. I used them as
-test cases when working out the typechecking rules, but they're not integrated
-into the rest of the text.
-
-A:
-
-```
-// OK
-fn F[Ts:! [Type;]](..., [:]args: Ts) -> i32;
+##### Tuple type
 
-fn G[U:! Type; Us:! [Type;]](arg: Vector(U), ..., [:]args: Vector(Us)) -> i32 {
-  return F(arg, ..., args);
-}
-```
+FIXME too much nesting
 
-B:
+At type checking time, we don't necessarily know which parameters each argument
+will match with, or the other way around. For example, consider the following
+code:
 
-```
-// OK (but maybe too complex to support?)
-fn F[Ts:! [Type;], Ts:! [Type;]](arg1: (..., [:]Ts), arg2: (..., [:]Ts)) -> i32;
+```carbon
+fun F(a: i32, ..., [:]b: [i32;], c: i32);
 
-fn G[Vs:! [Type;]](..., [:]args: Vector(Vs)) -> i32 {
-  return F((..., args), (..., args));
+fun G(..., [:]x: [i32;]) {
+  F(1, 2, ..., [:]x);
 }
 ```
 
-```
-fn F[Ts:! [Type;]](arg1: (..., [:]Ts), arg2: (..., [:]Ts)) -> i32;
+If `x` is empty, the `2` will match with `c`, and otherwise the `2` will match
+with an element of `b`. Similarly, if `x` is not empty, its last element will
+match `c`, and the remaining elements (if any) will match elements of `b`.
+However, at type-checking time we don't know the size of `x` yes, so we don't
+know which will occur. On the other hand, the `1` will always match `a`.
 
-fn G[Vs:! [Type;], Ws:! [Type;]](arg1: (..., [:]Vector(Vs)), arg2: (..., [:]Vector(Ws))) -> i32 {
-  // Error: Vs might be different from Ws.
-  return F(arg1, arg2);
-}
-```
+In general, We want type checking to fail if any possible monomorphization of
+the generic code would fail to typecheck. In this case that means we want type
+checking to fail if any of the potential argument-parameter mappings could fail
+to typecheck after monomorphization. Furthermore, for reasons of readability as
+well as efficiency, we want type checking to fail if any two potential mappings
+would deduce inconsistent values for any deduced parameter. However, in general
+this is intractable, because the number of distinct ways to map argument
+expressions to parameters is exponential in the number of variadic arguments.
```

Comment:
"exponential" here could use some justification (and I'm not sure it is precisely accurate).

Reply:
How's this?

I don't want to get too into the weeds here, because it feels pretty intuitive to me that anything that starts with "the number of distinct ways" probably isn't going to have a happy ending

Reply:
Maybe it would be better to avoid "exponential" as too specific, and use the more generic https://en.wikipedia.org/wiki/Combinatorial_explosion term instead?

Reply:
I'm not sure how to use "combinatorial explosion" as an adjective, especially if I need to provide some justification.